

<!DOCTYPE html>
<html lang="">
<head prefix="og: http://ogp.me/ns#">
    <!-- hexo-inject:begin --><!-- hexo-inject:end --><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=1">
    <meta name="description" content="">
    <title>Needleman–Wunsch algorithm | Lingkun&#39;s Small House</title>

    <!-- Open Graph -->
    <meta name="description" content="IntroductionThe Needleman&amp;#x2013;Wunsch algorithm is an algorithm used in bioinformatics to align protein or nucleotide sequences. It was one of the first applications of dynamic programming to compar">
<meta name="keywords" content="code">
<meta property="og:type" content="article">
<meta property="og:title" content="Needleman–Wunsch algorithm">
<meta property="og:url" content="https://ohyoukillkenny.github.io/blog/02/06/DNA-min-mismatch/index.html">
<meta property="og:site_name" content="Lingkun&#39;s Small House">
<meta property="og:description" content="IntroductionThe Needleman&amp;#x2013;Wunsch algorithm is an algorithm used in bioinformatics to align protein or nucleotide sequences. It was one of the first applications of dynamic programming to compar">
<meta property="og:locale" content="default">
<meta property="og:updated_time" content="2018-02-06T03:16:49.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Needleman–Wunsch algorithm">
<meta name="twitter:description" content="IntroductionThe Needleman&amp;#x2013;Wunsch algorithm is an algorithm used in bioinformatics to align protein or nucleotide sequences. It was one of the first applications of dynamic programming to compar">

    <script type="application/ld+json">
{
  "@context": "http://schema.org",
  "@type": "Person",
  "email": "mailto:",
  "image": "",
  "name": "Lingkun Kong",
  "url": "https://ohyoukillkenny.github.io/blog"
}
</script>
    <script>
    var algoliaEnabled = false;
    
</script>

    

    
    
        <link href="https://fonts.googleapis.com/css?family=Source+Code+Pro" rel="stylesheet" type="text/css">
    
    
    
        <!-- stylesheets list from config.yml -->
        
        <link rel="stylesheet" href="/blog/css/aloha.css">
        
        <link rel="stylesheet" href="//cdn.bootcss.com/semantic-ui/2.2.4/semantic.min.css">
        
        <link rel="stylesheet" href="//cdn.bootcss.com/magnific-popup.js/1.1.0/magnific-popup.min.css"><!-- hexo-inject:begin --><!-- hexo-inject:end -->
        
    

</head>
<body id="body" style="margin-bottom: 0;" class="pushable">

    <!-- hexo-inject:begin --><!-- hexo-inject:end --><div class="ui top fixed menu">
        <a class="item" id="menu-icon"><i class="sidebar icon"></i></a>
    </div>

    <div id="menu-sidebar" class="ui left vertical sidebar menu">

    <div id="sidebar-top">
        <div class="content">
            <h3>Lingkun&#39;s Small House</h3>
        </div>
    </div>
    <div class="ui container sidebar-card">
    <div class="ui people shape content">
        <div class="active side">
            <div class="ui card">
                <div class="image">
                    <img class="ui medium bordered image" src="/blog/images/avatar.jpg">
                </div>
                <div class="content">
                    <a class="header">Lingkun Kong</a>
                    
                        <div class="meta">
                            Programmer / Basketball Fan / Pingpong Player
                        </div>
                    
                </div>
                <div class="extra content">
                    <div class="ui list">
                        
                        <div class="item">
                            <i class="marker icon" style="float: left"></i>
                            <div class="content">Shanghai</div>
                        </div>
                        
                        
                    </div>

                </div>
                <div class="extra content">
                    <div class="ui list">
                        

                                <a href="https://github.com/Ohyoukillkenny"><i class="large github icon" style="float: left"></i></a>

                            
                    </div>
                </div>
            </div>
        </div>
    </div>


</div>

    
    

    

    
    <a href="/index.html" class="item" >
        <i class="home icon"></i>
        HOME
    </a>

    

    
</div>

    <div class="pusher body-content">
        <div id="content" class="ui main container">
            <!--<div class="ui one column grid">-->
                <!--<div class="column">-->
                    <!--<div class="ui main container">-->
                        <div id="post-DNA-min-mismatch" class="ui main container article-type-post">

    

    <div class="ui divided grid">

        <div class="two column row" id="article-content">

            <div class="thirteen wide computer thirteen wide tablet sixteen wide mobile column">
                

<h1 class="ui header">
    
    Needleman–Wunsch algorithm
</h1>



                <div class="article-inner">

                    <div class="article-entry" itemprop="articleBody">
                        
                        <h3 id="Introduction"><a href="#Introduction" class="headerlink" title="Introduction"></a>Introduction</h3><p>The <strong>Needleman&#x2013;Wunsch algorithm</strong> is an <a href="https://en.wikipedia.org/wiki/Algorithm" target="_blank" rel="noopener">algorithm</a> used in <a href="https://en.wikipedia.org/wiki/Bioinformatics" target="_blank" rel="noopener">bioinformatics</a> to <a href="https://en.wikipedia.org/wiki/Sequence_alignment" target="_blank" rel="noopener">align</a> <a href="https://en.wikipedia.org/wiki/Protein" target="_blank" rel="noopener">protein</a> or <a href="https://en.wikipedia.org/wiki/Nucleotide" target="_blank" rel="noopener">nucleotide</a> sequences. It was one of the first applications of <a href="https://en.wikipedia.org/wiki/Dynamic_programming" target="_blank" rel="noopener">dynamic programming</a> to compare biological sequences.</p>
<p>In this article, we try to use Needleman-Wunsch algorithm to  find the minimum mismatch between two different DNA sequences.</p>
<h4 id="Definition-of-Mismatches"><a href="#Definition-of-Mismatches" class="headerlink" title="Definition of Mismatches"></a>Definition of Mismatches</h4><p>Here we have two different DNA sequences, i.e., <code>S1</code> and <code>S2</code>, where</p>
<p><code>S1 = &quot;GCATGCU&quot;</code>,</p>
<p><code>S2 = &quot;GCTTAGCU&quot;</code>.</p>
<p>One solution to match these two DNA sequences is listed as below:</p>
<p><code>S1* = &quot;GCAT-GCU&quot;</code>,</p>
<p><code>S2* = &quot;GCTTAGCU&quot;</code>,</p>
<p>in which <code>-</code> is called as &#x201C;gap&#x201D;, i.e., in this position the DNA is supposed to be unknown.</p>
<p><strong>Match</strong>: The two letters are the same, except both two letters are gapped.</p>
<p><strong>Mismatch</strong>: The two letters are differential or both two letters are gapped.</p>
<p>By above definition, the solution we find for matching <code>S1</code> and <code>S2</code> has two mismatches &#x2014; <code>T</code> &amp; <code>A</code> and <code>-</code> &amp; <code>A</code>.</p>
<h4 id="Codes-for-finding-minimum-mismatches"><a href="#Codes-for-finding-minimum-mismatches" class="headerlink" title="Codes for finding minimum mismatches"></a>Codes for finding minimum mismatches</h4><p>Here I list the python code for finding minimum number of mismatches between two DNA sequences. The insights of codes below can be viewed by supplemental documents <a href="./insights.pdf">here</a>.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="string">input:</span></span><br><span class="line"><span class="string">    S1 = [&apos;A&apos;,&apos;B&apos;,&apos;C&apos;,&apos;D&apos;]</span></span><br><span class="line"><span class="string">    S2 = [&apos;A&apos;,&apos;B&apos;,&apos;C&apos;]</span></span><br><span class="line"><span class="string">output:</span></span><br><span class="line"><span class="string">    1</span></span><br><span class="line"><span class="string">print info:</span></span><br><span class="line"><span class="string">    New S1 is:   ABCD</span></span><br><span class="line"><span class="string">    New S2 is:   ABC-</span></span><br><span class="line"><span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">MinMismatch</span><span class="params">(S1, S2)</span>:</span></span><br><span class="line">    <span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="string">    First of all, we construct matrix for counting maximum length of common subsequence.</span></span><br><span class="line"><span class="string">    &apos;&apos;&apos;</span></span><br><span class="line">    l1, l2 = len(S1), len(S2)</span><br><span class="line">    matrix = [[<span class="number">0</span> <span class="keyword">for</span> j <span class="keyword">in</span> range(l2+<span class="number">1</span>)] <span class="keyword">for</span> i <span class="keyword">in</span> range(l1+<span class="number">1</span>)]</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>, l1+<span class="number">1</span>):</span><br><span class="line">        <span class="keyword">for</span> j <span class="keyword">in</span> range(<span class="number">1</span>, l2+<span class="number">1</span>):</span><br><span class="line">            <span class="keyword">if</span> S1[i<span class="number">-1</span>] == S2[j<span class="number">-1</span>]:</span><br><span class="line">                matrix[i][j] = matrix[i<span class="number">-1</span>][j<span class="number">-1</span>] + <span class="number">1</span></span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                matrix[i][j] = max(matrix[i<span class="number">-1</span>][j<span class="number">-1</span>], matrix[i][j<span class="number">-1</span>], matrix[i<span class="number">-1</span>][j])</span><br><span class="line">    <span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="string">    Dynamic programming. Tracing back the matrix and construct two DNA sequences with gaps.</span></span><br><span class="line"><span class="string">    We saved the restructured sequences into A (originally S1) and B (originally S2).</span></span><br><span class="line"><span class="string">    &apos;&apos;&apos;</span></span><br><span class="line">    left_up, up, left = <span class="number">0</span>, <span class="number">1</span>, <span class="number">2</span></span><br><span class="line">    i, j = l1, l2</span><br><span class="line">    A, B = [], []</span><br><span class="line">    <span class="keyword">while</span> (i,j) != (<span class="number">0</span>,<span class="number">0</span>):</span><br><span class="line">        <span class="keyword">if</span> i == <span class="number">0</span>:</span><br><span class="line">            i, j = i, j<span class="number">-1</span></span><br><span class="line">            flag = left</span><br><span class="line">        <span class="keyword">elif</span> j == <span class="number">0</span>:</span><br><span class="line">            i, j = i<span class="number">-1</span>, j</span><br><span class="line">            flag = up</span><br><span class="line">        <span class="keyword">elif</span> S1[i<span class="number">-1</span>] == S2[j<span class="number">-1</span>] <span class="keyword">or</span> matrix[i<span class="number">-1</span>][j<span class="number">-1</span>] == max(matrix[i<span class="number">-1</span>][j<span class="number">-1</span>], matrix[i][j<span class="number">-1</span>], matrix[i<span class="number">-1</span>][j]):</span><br><span class="line">            i, j, flag = i<span class="number">-1</span>, j<span class="number">-1</span>, left_up</span><br><span class="line">        <span class="keyword">elif</span> matrix[i<span class="number">-1</span>][j] == max(matrix[i<span class="number">-1</span>][j<span class="number">-1</span>], matrix[i][j<span class="number">-1</span>], matrix[i<span class="number">-1</span>][j]):</span><br><span class="line">            i, j, flag = i<span class="number">-1</span>, j, up</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            i, j, flag = i, j<span class="number">-1</span>, left</span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> flag == left_up:</span><br><span class="line">            A.append(S1[i])</span><br><span class="line">            B.append(S2[j])</span><br><span class="line">        <span class="keyword">elif</span> flag == up:</span><br><span class="line">            A.append(S1[i])</span><br><span class="line">            B.append(<span class="string">&apos;-&apos;</span>)</span><br><span class="line">        <span class="keyword">else</span>:</span><br><span class="line">            A.append(<span class="string">&apos;-&apos;</span>)</span><br><span class="line">            B.append(S2[j])</span><br><span class="line">    <span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="string">    Print info about new DNA sequences with minimum number of mismatches.</span></span><br><span class="line"><span class="string">    &apos;&apos;&apos;</span></span><br><span class="line">    A.reverse()</span><br><span class="line">    B.reverse()</span><br><span class="line">    printA, printB = <span class="string">&apos;&apos;</span>,<span class="string">&apos;&apos;</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(len(A)):</span><br><span class="line">        printA += A[i]</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(len(B)):</span><br><span class="line">        printB += B[i]</span><br><span class="line">    print(<span class="string">&apos;New S1 is:   &apos;</span>+ printA)</span><br><span class="line">    print(<span class="string">&apos;New S2 is:   &apos;</span>+ printB)</span><br><span class="line">    <span class="string">&apos;&apos;&apos;</span></span><br><span class="line"><span class="string">    Calculate the number of mismatches.</span></span><br><span class="line"><span class="string">    &apos;&apos;&apos;</span></span><br><span class="line">    mismatch = <span class="number">0</span></span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> range(len(A)):</span><br><span class="line">        <span class="keyword">if</span> A[i] != B[i] <span class="keyword">or</span> (A[i] == <span class="string">&apos;-&apos;</span> <span class="keyword">and</span> B[i] == <span class="string">&apos;-&apos;</span>):</span><br><span class="line">            mismatch += <span class="number">1</span></span><br><span class="line">    <span class="keyword">return</span> mismatch</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__==<span class="string">&quot;__main__&quot;</span>:</span><br><span class="line">    S1 = list(<span class="string">&quot;GGATCGA&quot;</span>)</span><br><span class="line">    S2 = list(<span class="string">&quot;GAATTCAGTTA&quot;</span>)</span><br><span class="line">    print(MinMismatch(S1,S2))</span><br></pre></td></tr></table></figure>

                        
                    </div>

                    
                    
                    

                    <div class="description post-description">
    <span class="post-description-item">
        <time datetime="2018-02-06T02:47:55.000Z" itemprop="datePublished">2018-02-06</time>

    </span>
    <span class="post-description-item">
        

    <span>
        <i class="tags icon"></i>
    </span>

    
    <span class="ui tiny label">
        <a href="/blog/tags/code/">code</a>
    </span>

    

    </span>
</div>

                </div>
                
                <div id="pagination" class="nav-web ui text container pagination">

    <div class="ui stackable two column divided grid container">
        <div class="row">
            <div class="column nav-left">
                

                <a href="/blog/02/06/hanoi/" class="ui tiny button">

                    <i class="angle left icon"></i>
                    
                    Hanoi in Assembly x86 (NASM)
                    

                </a>

                
            </div>
            <div class="column nav-right">
                

                <a href="/blog/01/08/micro-payment/" class="ui tiny button">
                    
                    Micropayment
                    
                    <i class="angle right icon"></i></a>
                
            </div>
        </div>
    </div>

</div>
                
            </div>

            
            <div class="three wide computer three wide tablet column">

                <div class="ui sticky article-toc" id="article-toc">
                    <!-- toc -->
                    <ol class="ui ordered list"><li><a href="#Introduction"><span>Introduction</span></a><ol><li><a href="#Definition-of-Mismatches"><span>Definition of Mismatches</span></a></li><li><a href="#Codes-for-finding-minimum-mismatches"><span>Codes for finding minimum mismatches</span></a></li></ol></li></ol>
                </div>

            </div>
            

        </div>
    </div>
</div>


                    <!--</div>-->
                <!--</div>-->
            <!--</div>-->
        </div>
        <div class="ui vertical footer segment">
    <div class="ui center aligned container">
        <div class="ui inverted section divider"></div>
        &copy; 2018 <a href="/">Lingkun Kong</a>,
        Powered by <a href="https://github.com/henryhuang/hexo-theme-aloha" target="_blank">Aloha</a> and <a
                href="http://hexo.io/" target="_blank">Hexo</a>.
    </div>
</div>

    </div>




<!-- scripts list from theme config.yml -->

<script src="//cdn.bootcss.com/jquery/3.1.0/jquery.min.js"></script>

<script src="//cdn.bootcss.com/semantic-ui/2.2.4/semantic.min.js"></script>

<script src="//cdn.bootcss.com/algoliasearch/3.18.1/algoliasearch.min.js"></script>

<script src="//cdn.bootcss.com/algoliasearch-helper-js/2.13.0/algoliasearch.helper.min.js"></script>

<script src="/blog/js/semantic-ui-algolia.js"></script>

<script src="/blog/js/aloha-events.js"></script><!-- hexo-inject:begin --><!-- hexo-inject:end -->



</body>
</html>
